### PR TITLE
Added PyTorch fcnresnet101 segmentation conversion cases 

### PR DESCRIPTION
**Pull request by richard28039**
We write a sample code about transforming Pytorch fcnresnet101 to ONNX running on OpenCV.

The input source image was shooted by ourself.

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [X] I agree to contribute to the project under Apache 2 License.
- [X] To the best of my knowledge, the proposed patch is not based on a code under GPL or another license that is incompatible with OpenCV
- [X] The PR is proposed to the proper branch
- [ ] There is a reference to the original bug report and related work
- [ ] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [ ] The feature is well documented and sample code can be built with the project CMake

### CODE REVIEW

File: samples/dnn/dnn_model_runner/dnn_conversion/pytorch/segmantation/pytorch_fcnresnet101.py [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+    # Save the result image
+    cv2.imwrite(save_path, result)
+    print(f'Success predict the image at {save_path}')
+
+def get_opencv_dnn_prediction(opencv_net, original_image, preproc_img):
+    # set OpenCV DNN input
+    opencv_net.setInput(preproc_img)
+
+    # OpenCV DNN inference
+    predict = opencv_net.forward()
```
Comment (dkurt) [inline samples/dnn/dnn_model_runner/dnn_conversion/pytorch/segmantation/pytorch_fcnresnet101.py:107]: Why not to use existing sample https://github.com/opencv/opencv/blob/4.x/samples/dnn/segmentation.py?

File: samples/dnn/dnn_model_runner/dnn_conversion/pytorch/segmantation/data/cat_and_bike.jpg
Comment (dkurt) [inline samples/dnn/dnn_model_runner/dnn_conversion/pytorch/segmantation/data/cat_and_bike.jpg:1]: Do not add image files to the repo please

Comment (richard28039) [reply samples/dnn/dnn_model_runner/dnn_conversion/pytorch/segmantation/data/cat_and_bike.jpg:1]: Hello sir, we had removed the image files from the repo and change to use argparse method get input picture.

File: samples/dnn/dnn_model_runner/dnn_conversion/pytorch/segmantation/pytorch_fcnresnet101.py (hunk 1 above)
Comment (richard28039) [reply samples/dnn/dnn_model_runner/dnn_conversion/pytorch/segmantation/pytorch_fcnresnet101.py:108]: Hello sir, we tried to use the existing sample but an error occurred. It seemed that it did not support ONNX input, so we wrote another sample code convert other model to the ONNX model. This is our idea about the work. Thank for your review.
error

Comment (dkurt) [reply samples/dnn/dnn_model_runner/dnn_conversion/pytorch/segmantation/pytorch_fcnresnet101.py:108]: May I ask to refactor the sample in such case? `--framework` can be removed completely and `onnx` should be `args.model`:
```
net = cv.dnn.readNet(args.model, args.config)
```

Comment (richard28039) [reply samples/dnn/dnn_model_runner/dnn_conversion/pytorch/segmantation/pytorch_fcnresnet101.py:108]: sure, would keep you posted.

File: samples/dnn/models.yml [hunk 2]
```diff
@@ -178,3 +178,18 @@ fcn8s:
...
   sample: "segmentation"
+
+fcnresnet101:
+  load_info:
+    url: "https://github.com/richard28039/fcnresnet101_onnx_model/raw/main/fcn_resnet101.onnx?download="
+    sha1: "1242104e22c9eb0b320d7ae36b11cdd1fda40857"
+  model: "fcn_resnet101.onnx"
+  mean: [103.5, 116.2, 123.6]
+  std : [0.225, 0.224, 0.229]
+  scale: 0.00392
```
Comment (dkurt) [inline samples/dnn/models.yml:189]: recompute mean/scale considering std parameter.
```
blob = [(inp - mean) * scale / std]
```
is the same as
```
```
blob = [(inp - mean / std) * (scale / std)]
```

Comment (richard28039) [reply samples/dnn/models.yml]: Hello sir, I had apply your formula on my parameter, now the onnx model can run on the sample code, thank you.

File: samples/dnn/segmentation.py [hunk 3]
```diff
@@ -102,7 +102,7 @@ def showLegend(classes):
     inpWidth = args.width if args.width else frameWidth
     inpHeight = args.height if args.height else frameHeight
     blob = cv.dnn.blobFromImage(frame, args.scale, (inpWidth, inpHeight), args.mean, args.rgb, crop=False)
-
+
```
Comment (dkurt) [inline samples/dnn/segmentation.py:105]: Please remove trailing witespaces

Comment (richard28039) [reply samples/dnn/segmentation.py]: Hello sir, I had removed the whitespace, thank you!

File: samples/dnn/models.yml [hunk 4]
```diff
@@ -178,3 +178,16 @@ fcn8s:
   height: 500
   rgb: false
   sample: "segmentation"
+
+fcnresnet101:
+  load_info:
+    url: "https://github.com/richard28039/fcnresnet101_onnx_model/raw/main/fcn_resnet101.onnx?download="
```
Comment (dkurt) [inline samples/dnn/models.yml:184]: Where we can find an original model?

Comment (richard28039) [reply samples/dnn/models.yml:184]: Hello sir, the model and clolor file are here https://github.com/richard28039/fcnresnet101_onnx_model

File: samples/dnn/models.yml [hunk 5]
```diff
@@ -178,3 +178,16 @@ fcn8s:
...
+    sha1: "1242104e22c9eb0b320d7ae36b11cdd1fda40857"
+  model: "fcn_resnet101.onnx"
+  mean: [103.5, 116.2, 123.6]
+  scale: 0.019
+  width: 500
+  height: 500
+  rgb: false
+  classes: "pascal-voc-classes.txt"
```
Comment (dkurt) [inline samples/dnn/models.yml:192]: `pascal-voc-classes.txt` missed file

File: samples/dnn/models.yml (hunk 4 above)
Comment (dkurt) [reply samples/dnn/models.yml:184]: This is a newly created repository. Can you please refer how ONNX model was created?

File: samples/dnn/models.yml (hunk 5 above)
Comment (richard28039) [reply samples/dnn/models.yml:192]: Hello sir, classes file is here https://github.com/richard28039/fcnresnet101_onnx_model/blob/main/pascal-voc-classes.txt

File: samples/dnn/models.yml (hunk 4 above)
Comment (richard28039) [reply samples/dnn/models.yml:184]: Hello sir, I added the create code in my repository.

File: samples/dnn/models.yml (hunk 5 above)
Comment (dkurt) [reply samples/dnn/models.yml:192]: Please use https://github.com/opencv/opencv/blob/4.x/samples/data/dnn/object_detection_classes_pascal_voc.txt instead. They match each other.

File: samples/dnn/models.yml (hunk 4 above)
Comment (dkurt) [reply samples/dnn/models.yml:184]: Is there any difference with https://github.com/onnx/models/blob/main/vision/object_detection_segmentation/fcn/model/fcn-resnet101-11.onnx ?

Comment (richard28039) [reply samples/dnn/models.yml:184]: I had try this model berfore and it occurred the problem below, it seem that cv2.dnn.readnet can not accept dynamic shape, so I create the model which shape is (1, 3, 500, 500) by myself.
螢幕擷取畫面 2023-10-19 181108

File: samples/dnn/models.yml [hunk 6]
```diff
@@ -178,3 +178,16 @@ fcn8s:
...
+  load_info:
+    url: "https://github.com/richard28039/fcnresnet101_onnx_model/raw/main/fcn_resnet101.onnx?download="
+    sha1: "1242104e22c9eb0b320d7ae36b11cdd1fda40857"
+  model: "fcn_resnet101.onnx"
+  mean: [103.5, 116.2, 123.6]
+  scale: 0.019
+  width: 500
+  height: 500
+  rgb: false
+  classes: "object_detection_classes_pascal_voc.txt"
```
Comment (dkurt) [inline samples/dnn/models.yml:192]: Please add an image with resulting segmentation from a sample launch. I have a concern that classes might be 1 position shifted (depends on background id).

So please run a sample on an image with more that one segmented object and put to PR description or in this thread,

Comment (richard28039) [reply samples/dnn/models.yml:192]: Hello @dkurt , I think your concern is right, the background classes might have to be add.
Should I add background class to https://github.com/opencv/opencv/blob/4.x/samples/data/dnn/object_detection_classes_pascal_voc.txt ?

without background
image
image

with background
image
image

File: samples/dnn/models.yml (hunk 4 above)
Comment (dkurt) [reply samples/dnn/models.yml:184]: Thanks, I will try to fix it

Comment (dkurt) [reply samples/dnn/models.yml:184]: https://github.com/opencv/opencv/pull/24463 going to fix problem with dynamic shapes. So please replace url to https://github.com/onnx/models/raw/fb8271d5d5d9b90dbb1eb5e8e40f8f580fb248b3/vision/object_detection_segmentation/fcn/model/fcn-resnet101-11.onnx

Comment (richard28039) [reply samples/dnn/models.yml:184]: OK, I will replace url, thanks.

File: samples/dnn/models.yml (hunk 6 above)
Comment (dkurt) [reply samples/dnn/models.yml:192]: Thanks, problem should be solved in a common.py script which will affect other samples so I'd like to merge this PR as is.

**Review by dkurt — APPROVED** [review]

File: samples/dnn/models.yml (hunk 6 above)
Comment (richard28039) [reply samples/dnn/models.yml:192]: I got it, thank you!

**Review by asmorkalov — APPROVED** [review]
:+1:

**Review by asmorkalov — COMMENTED** [review]
```
python3 ./segmentation.py fcnresnet101
[ WARN:0@0.019] global samples.cpp:61 findFile cv::samples::findFile('object_detection_classes_pascal_voc.txt') => ''
[ERROR:0@0.332] global onnx_importer.cpp:2441 parseShape DNN/ONNX(Shape): dynamic 'zero' shapes are not supported, input input [ 0 3 0 0 ]
[ERROR:0@0.332] global onnx_importer.cpp:1033 handleNode DNN/ONNX: ERROR during processing node with 1 inputs and 1 outputs: [Shape]:(onnx_node!Shape_0) from domain='ai.onnx'
Traceback (most recent call last):
  File "./segmentation.py", line 83, in <module>
    net = cv.dnn.readNet(args.model, args.config, args.framework)
cv2.error: OpenCV(4.8.0-dev) /home/alexander/Projects/OpenCV/opencv-master/modules/dnn/src/onnx/onnx_importer.cpp:1052: error: (-2:Unspecified error) in function 'handleNode'
> Node [Shape@ai.onnx]:(onnx_node!Shape_0) parse error: OpenCV(4.8.0-dev) /home/alexander/Projects/OpenCV/opencv-master/modules/dnn/src/onnx/onnx_importer.cpp:2442: error: (-215:Assertion failed) !isDynamicShape in function 'parseShape'
> 
```
What do I do wrong?

**Comment by asmorkalov** [comment]
With related patch from @dkurt:
```
[ WARN:0@0.019] global samples.cpp:61 findFile cv::samples::findFile('object_detection_classes_pascal_voc.txt') => ''
[ERROR:0@0.329] global onnx_importer.cpp:1033 handleNode DNN/ONNX: ERROR during processing node with 1 inputs and 1 outputs: [MaxPool]:(onnx_node!MaxPool_8) from domain='ai.onnx'
Traceback (most recent call last):
  File "./segmentation.py", line 83, in <module>
    net = cv.dnn.readNet(args.model, args.config, args.framework)
cv2.error: OpenCV(4.8.0-dev) /home/alexander/Projects/OpenCV/opencv-master/modules/dnn/src/onnx/onnx_importer.cpp:1052: error: (-2:Unspecified error) in function 'handleNode'
> Node [MaxPool@ai.onnx]:(onnx_node!MaxPool_8) parse error: OpenCV(4.8.0-dev) /home/alexander/Projects/OpenCV/opencv-master/modules/dnn/src/layers/pooling_layer.cpp:1338: error: (-215:Assertion failed) inputs.size() != 0 in function 'getMemoryShapes'
> 
```

**Comment by dkurt** [comment]
@asmorkalo, can confirm, let me check.

**Comment by asmorkalov** [comment]
I can confirm that fix from @dkurt works. The sample works properly, if I add "background" to the beggining of `object_detection_classes_pascal_voc.txt`. Otherwise I get exception
```
alexander@paradox:~/Projects/OpenCV/opencv-master/samples/dnn$ python3 ./segmentation.py fcnresnet101
Traceback (most recent call last):
  File "./segmentation.py", line 133, in <module>
    showLegend(classes)
  File "./segmentation.py", line 70, in showLegend
    assert(len(classes) == len(colors))
```
I propose to force add background in the sample code if len(classes) - len(colors) = 1

**Comment by richard28039** [comment]
@dkurt on the background question about object_detection_classes_pascal_voc.txt , the problem should be solved in a common.py in another pr as you originally said, or should I modify it in the way that @asmorkalov propose, thank your!

**Comment by asmorkalov** [comment]
I propose to add the hack to current PR, because the sample just crashed in default configuration. Python part may be modified in a separate PR.

**Comment by dkurt** [comment]
@richard28039, there are two ways to avoid crash: 1) temporary add "backdround" label to the classes array 2) remove `classes: "object_detection_classes_pascal_voc.txt"` for `fcnresnet101` from models.yml - this way sample just colorize the segmentation without a legend. I prefer second option.

**Comment by richard28039** [comment]
Hello, I removed the `classes` in `fcnresnet101` from `models.yml` and the result looks good, thanks for your suggestions.

**Pull request merged**